Add AWS ACM permissions

matt-telstra opened the pull request:
For running tests for [PR 60552](https://github.com/ansible/ansible/pull/60552).

s-hertel commented on aws/policy/security-services.yaml:
```diff
@@ -110,3 +110,20 @@ Statement:
+      - acm:DeleteCertificate
+    Resource:
+      - arn:aws:acm:us-east-1:{{ aws_account_id }}:certificate/*
+  - Sid: AllowACMRestrictable
```
Looks like these two could go in the `AllowGlobalUnrestrictedResourceActionsWhichIncurNoFees` Sid instead.

matt-telstra replied:
Do you want me to merge `AllowACMRestrictable` into `AllowGlobalResourceRestrictedActionsWhichIncurNoFees`? Why? Isn't it better to have more fine grained permissions, rather than allow the `acm:*` actions on `arn:aws:iam::{{ aws_account_id }}:role/*`, and the other way around?

matt-telstra commented:
I don't understand how to create a Terminator class, nor why.

My test playbook wraps everything up in a `block`, and has an `always` statement to remove resources if the test fails.

When I look at the existing code in `security_services.py` I see stuff like `@abc.abstractmethod`, and I don't quite understand what that's supposed to do.

* What is a "Terminator" class? 
* Why do some have `create` functions. Creating what? This class, or the resource? 
* How to I specify which boto client to use? 
* Is `terminate` the function I need to populate? 
* Can I leave out the `@properties`? Which `@properties` do I need?
* What does the README mean when it says "if resources have a created time"? Don't all AWS resources have a created timestamp?
* Should I be able to follow the instructions in that README if I don't have the keys to your account?

s-hertel commented:
Your tests look good and you've done all the precautions there with `always` etc right, but there's always the possibility something could go wrong - either with shippable, or an AWS service issue, etc. We need an automated way to find and clean up these potential leaked resources in our account. Our solution is to deploy a lambda function that runs these terminator classes.

We terminate resources based on their age, and because not all resources return a CreatedTime timestamp or something similar (really, AWS is wildly inconsistent at times), those without use a database-backed base class so we can store them with the time when the terminator first finds them and then use that to approximate how old they and when they need to be cleaned up.

* If your resource has a timestamp of creation associated use the Terminator base class
* If your resource does not, use the DbTerminator base class

You should have the properties `create`, `id`, `name`, and `terminate`. If there's a created time that's returned when you list a resource type you should also have `created_time`. You can fine tune some things by adding `ignore`/`age_limit`, but those are special cases and not relevant for this one.

The `create` method returns the credentials to create the client, the class name, the boto3 resource name to create the client, and a function for the client to use which will list all the given resources for that resource type. That function should return a list of dictionaries, where each dictionary describes a single resource, in this case, a certificate. When you see `self.instance` used, that is a single dictionary from that list, describing one certificate.

In `terminate` you can use self.client to use the same client.

You could modify the accounts used in aws/config.yml so you can test by using the aws/cleanup.py script, but I don't think it's necessary. Generally the code is pretty copy/pastable and just needs to be tweaked for the new resource. I'll check things are working before I merge your pull request. I'm sure you know the acm service better than me, so take that into consideration, but I think this code should do it (haven't checked for syntax errors/typos):
```
class ACMCertificate(DbTerminator):
    @staticmethod
    def create(credentials):
        def get_paginated_certificates(client):
            return client.get_paginator('list_certificates').paginate().build_full_result()['CertificateSummaryList']
        return Terminator._create(credentials, ACMCertificate, 'acm', get_paginated_certificates)

    @property
    def id(self):
        return self.instance['CertificateArn']

    @property
    def name(self):
        return self.instance['DomainName']

    def terminate(self):
        self.client.delete_certificate(CertificateArn=self.id)
```
However, since it looks like `client.describe_certificate` does return a timestamp, you could write it as:
```
class ACMCertificate(Terminator):
    @staticmethod
    def create(credentials):
        def get_paginated_certificates(client):
            return client.get_paginator('list_certificates').paginate().build_full_result()['CertificateSummaryList']

        def get_detailed_certificates(client):
            detailed_certs = []
            for certificate in get_paginated_certificates(client):
                detailed_certs.append(client.describe_certificate(CertificateArn=certificate['CertificateArn'])['Certificate']
            return detailed_certs

        return Terminator._create(credentials, ACMCertificate, 'acm', get_detailed_certificates)

    @property
    def created_time(self):
        return self.instance['CreatedAt']

    @property
    def id(self):
        return self.instance['CertificateArn']

    @property
    def name(self):
        return self.instance['DomainName']

    def terminate(self):
        self.client.delete_certificate(CertificateArn=self.id)
```

If you have suggestions on the README, please feel free to file an issue with some constructive criticism or a pull request to improve it. I wrote it initially for internal use before this repo was opened and I'd like to make it more useful for contributors in general because it's great that people can get their tests running in CI without waiting on a core dev.

matt-telstra commented:
I've added some comments to that PR. 

I've added a new terminator class. The differences from what you wrote were:

* There was a missing bracket
* The `CreatedAt` attribute is only valid for AWS provided certs. My new module is for imported certs, which use `ImportedAt`. I accounted for both in the terminator class, in case someone adds functionality for AWS provided certs later.

matt-telstra commented on aws/policy/security-services.yaml:
```diff
@@ -93,7 +93,14 @@ Statement:
       - inspector:DescribeAssessmentTemplates
       - inspector:ListTagsForResource
       - inspector:ListEventSubscriptions
+      - acm:ImportCertificate
```
Hmm. Odd.

When I try to create a new policy in the console with just `acm:ListCertificates`  it won't let me restrict the resource. 

> The action you chose supports all resources

When I try to create a new policy with just `acm:ImportCertificate`, it lets me restrict the resource.

Oh well. If this commit fixes the problem for you, that's good.

jillr replied:
I've tested with:
```
  - Sid: AllowACMRestrictable
    Effect: Allow
    Action:
      - acm:DescribeCertificate
      - acm:GetCertificate
      - acm:AddTagsToCertificate
      - acm:DeleteCertificate
      - acm:ImportCertificate
      - acm:ListCertificates
```
(so removing both List and Import from `AllowGlobalUnrestrictedResourceActionsWhichIncurNoFees` and placing them both in `AllowACMRestrictable`)
The integration tests then seem to work fine - except we've hit a quota on the CI account so we'll need to get that sorted with AWS support before we can test more extensively.
